File: src/typescript/load.js
//File to store the load operations - in typescript
//Project structure: When using other langauges, make new file of the same name.
//Ensure code in new file does the ~exact same~ thing as this file
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
//load data into a db
// import fs from 'fs';
// import path from 'path';
// import mysql from 'mysql2/promise';
// import dotenv from 'dotenv';
var fs = require('fs');
var path = require('path');
var mysql = require('mysql2/promise');
var dotenv = require('dotenv');
// Load environment variables
dotenv.config();
var loadJSONtoDB = function () { return __awaiter(_this, void 0, void 0, function () {
    var jsonFilePath, jsonData, records, connection, _i, records_1, record, query, values, error_1;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                jsonFilePath = path.resolve(__dirname, '../../data/transformed100ts.json');
                jsonData = fs.readFileSync(jsonFilePath, 'utf-8');
                records = JSON.parse(jsonData);
                return [4 /*yield*/, mysql.createConnection({
                        host: process.env.DB_HOST,
                        user: process.env.DB_USER,
                        password: process.env.DB_PASSWORD,
                        database: process.env.DB_NAME,
                    })];
            case 1:
                connection = _a.sent();
                _a.label = 2;
            case 2:
                _a.trys.push([2, 7, 8, 10]);
                _i = 0, records_1 = records;
                _a.label = 3;
            case 3:
                if (!(_i < records_1.length)) return [3 /*break*/, 6];
                record = records_1[_i];
                query = "\n        INSERT INTO TableNameTBD (\n          item_type, \n          order_priority, \n          order_date, \n          order_id, \n          units_sold, \n          unit_price, \n          unit_cost, \n          total_revenue, \n          total_cost, \n          total_profit\n        ) \n        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)\n      ";
                values = [
                    record["Item Type"],
                    record["Order Priority"],
                    record["Order Date"],
                    record["Order ID"],
                    record["Units Sold"],
                    record["Unit Price"],
                    record["Unit Cost"],
                    record["Total Revenue"],
                    record["Total Cost"],
                    record["Total Profit"],
                ];
                // Execute the query
                //Assumes no malicious SQL Injection
                return [4 /*yield*/, connection.execute(query, values)];
            case 4:
                // Execute the query
                //Assumes no malicious SQL Injection
                _a.sent();
                _a.label = 5;
            case 5:
                _i++;
                return [3 /*break*/, 3];
            case 6:
                console.log('Data loaded successfully!');
                return [3 /*break*/, 10];
            case 7:
                error_1 = _a.sent();
                console.error('Error loading data:', error_1);
                return [3 /*break*/, 10];
            case 8: return [4 /*yield*/, connection.end()];
            case 9:
                _a.sent();
                return [7 /*endfinally*/];
            case 10: return [2 /*return*/];
        }
    });
}); };
loadJSONtoDB().catch(console.error);
// CSV version
// const loadCSVtoDB = async () => {
//   // Step 1: Read the transformed CSV file
//   const csvFilePath = path.resolve(__dirname, 'transformedData.csv');
//   const csvData = fs.readFileSync(csvFilePath, 'utf-8');
//   // Parse CSV data (basic splitting; consider using a CSV parser library for complex files)
//   const rows = csvData.split('\n').map(row => row.split(','));
//   // Step 2: Connect to the AWS RDS MySQL/MariaDB database
//   const connection = await mysql.createConnection({
//     host: process.env.DB_HOST,
//     user: process.env.DB_USER,
//     password: process.env.DB_PASSWORD,
//     database: process.env.DB_NAME,
//   });
//   try {
//     // Step 3: Insert rows into the database
//     for (let row of rows) {
//       const query = `INSERT INTO your_table_name (column1, column2, ...) VALUES (?, ?, ...)`;
//       await connection.execute(query, row);
//     }
//     console.log('Data loaded successfully!');
//   } catch (error) {
//     console.error('Error loading data:', error);
//   } finally {
//     await connection.end();
//   }
// };
// loadCSVtoDB().catch(console.error);

File: src/java_awt/load.java
package java_awt;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.sql.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Load {

    public static void main(String[] args) {
        System.out.println("main");
        try {
            loadJSONtoDB();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void loadJSONtoDB() throws Exception {
        Class.forName("com.mysql.cj.jdbc.Driver");
        // Step 1: Load environment variables
        Map<String, String> properties = loadEnv(".env");
        String dbHost = properties.get("DB_HOST");
        String dbUser = properties.get("DB_USER");
        String dbPassword = properties.get("DB_PASSWORD");
        String dbName = properties.get("DB_NAME");
        String dbPort = properties.get("DB_PORT");
        System.out.println("host: " + dbHost);
        System.out.println("user: " + dbUser);
        System.out.println("pass: " + dbPassword);
        System.out.println("name: " + dbName);
        System.out.println("port: " + dbPort);

        // Step 2: Read the transformed JSON file
        String jsonFilePath = Paths.get("/tmp/transformed_100_TS.json").toAbsolutePath().toString();
        String jsonData = new String(Files.readAllBytes(Paths.get(jsonFilePath)));
        List<Map<String, Object>> records = parseJSON(jsonData);

        // Step 3: Connect to the AWS RDS MySQL/MariaDB database
        String jdbcUrl = "jdbc:mysql://tlq-pipeline-experiement-tcss462.cnm0wseaaqki.us-east-2.rds.amazonaws.com:3306/mydatabase";
        //+ dbHost + ":" + dbPort + "/" + dbName;
        System.out.println("MADE IT RIGHT BEFORE KNOWN ISSUE: " + jdbcUrl);
//        try (Connection connection = DriverManager.getConnection(jdbcUrl, dbUser, dbPassword)) {
        try (Connection connection = DriverManager.getConnection(jdbcUrl, dbUser, "")) {
            System.out.println("Connected to database successfully");


            // Step 4: Restart the database (Optional, use with caution)
           try (Statement statement = connection.createStatement()) {
               statement.execute("DROP DATABASE IF EXISTS db;");
               statement.execute("CREATE DATABASE IF NOT EXISTS db;");
               statement.execute("USE db;");

               String createTableQuery = """
                   CREATE TABLE IF NOT EXISTS sales (
                       item_type         VARCHAR(32)     NOT NULL,
                       order_priority    VARCHAR(1)      NOT NULL,
                       order_date        VARCHAR(32)     NOT NULL,
                       order_id          BIGINT          NOT NULL    PRIMARY KEY,
                       units_sold        BIGINT          NOT NULL,
                       unit_price        DECIMAL(10,2)   NOT NULL,
                       unit_cost         DECIMAL(10,2)   NOT NULL,
                       total_revenue     DECIMAL(10,2)   NOT NULL,
                       total_cost        DECIMAL(10,2)   NOT NULL,
                       total_profit      DECIMAL(10,2)   NOT NULL
                   );
               """;
               statement.execute(createTableQuery);
           }

           // Step 5: Insert records into the database
           String insertQuery = """
               INSERT INTO sales (
                   item_type,
                   order_priority,
                   order_date,
                   order_id,
                   units_sold,
                   unit_price,
                   unit_cost,
                   total_revenue,
                   total_cost,
                   total_profit
               ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
           """;

           try (PreparedStatement preparedStatement = connection.prepareStatement(insertQuery)) {
               for (Map<String, Object> record : records) {
                   preparedStatement.setString(1, (String) record.get("Item Type"));
                   preparedStatement.setString(2, (String) record.get("Order Priority"));
                   preparedStatement.setString(3, (String) record.get("Order Date"));
                   preparedStatement.setLong(4, ((Number) record.get("Order ID")).longValue());
                   preparedStatement.setLong(5, ((Number) record.get("Units Sold")).longValue());
                   preparedStatement.setBigDecimal(6, new java.math.BigDecimal(((Number) record.get("Unit Price")).doubleValue()));
                   preparedStatement.setBigDecimal(7, new java.math.BigDecimal(((Number) record.get("Unit Cost")).doubleValue()));
                   preparedStatement.setBigDecimal(8, new java.math.BigDecimal(((Number) record.get("Total Revenue")).doubleValue()));
                   preparedStatement.setBigDecimal(9, new java.math.BigDecimal(((Number) record.get("Total Cost")).doubleValue()));
                   preparedStatement.setBigDecimal(10, new java.math.BigDecimal(((Number) record.get("Total Profit")).doubleValue()));

                   preparedStatement.executeUpdate();
               }
           }
           System.out.println("Data loaded successfully!");
        } catch (Exception e) {
            System.out.println("ERRED OUT");
            System.err.println("Database Connection Error");
            System.err.println("Message: " + e.getMessage());
            System.err.println("Error loading data: " + e.getMessage());
            throw e;
        }
    }

    private static Map<String, String> loadEnv(String filePath) throws IOException {
        Map<String, String> properties = new HashMap<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (!line.trim().isEmpty() && line.contains("=")) {
                    String[] parts = line.split("=", 2);
                    properties.put(parts[0].trim(), parts[1].trim());
                }
            }
        }
        return properties;
    }

    private static List<Map<String, Object>> parseJSON(String jsonData) {
        List<Map<String, Object>> records = new ArrayList<>();
        jsonData = jsonData.trim();
        if (jsonData.startsWith("[") && jsonData.endsWith("]")) {
            jsonData = jsonData.substring(1, jsonData.length() - 1).trim();
            String[] objects = jsonData.split("},\\s*\\{");

            for (String object : objects) {
                object = object.replaceAll("[{}]", "").trim();
                String[] fields = object.split(",\s*");
                Map<String, Object> record = new HashMap<>();

                for (String field : fields) {
                    String[] keyValue = field.split(":", 2);
                    String key = keyValue[0].trim().replaceAll("\"", "");
                    String value = keyValue[1].trim().replaceAll("\"", "");
                    record.put(key, parseValue(value));
                }
                records.add(record);
            }
        }
        return records;
    }

    private static Object parseValue(String value) {
        if (value.matches("^-?\\d+\\.\\d+$")) {
            return Double.parseDouble(value);
        } else if (value.matches("^-?\\d+$")) {
            return Long.parseLong(value);
        } else {
            return value;
        }
    }
}
